### PR TITLE
[JENKINS-57111] Do not put agent offline for exceptions in onOnline

### PR DESCRIPTION
**Pull request by bitwiseman**
Agent implementers use onOnline to perform actions after an agent
put online.  The API specifies that exceptions thrown by this
method will not put an agent offline.  However, the default
implementation does exactly that.

This change clarifies the API to say all exceptions will be logged
and hardens setChannel against exceptions during onOnline.

NOTE: Errors will still cause an agent to go offline as they are
generally fatal conditions.

Also added tests to verify the behavior.
This change should address a range of open issues where onOnline throws an exception.

See [JENKINS-57111](https://issues.jenkins-ci.org/browse/JENKINS-57111).

### Proposed changelog entries

* Internal:  JENKINS-57111, Do not put agent offline for exceptions in onOnline().

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

### CODE REVIEW

**Review by oleg-nenashev — APPROVED** [review]
👍 https://javadoc.jenkins.io/hudson/slaves/ComputerListener.html#onOnline-hudson.model.Computer-hudson.model.TaskListener- is pretty specific about that:

```
Throws:
IOException - Exceptions will be recorded to the listener. Note that throwing an exception doesn't put the computer offline.
InterruptedException - Exceptions will be recorded to the listener. Note that throwing an exception doesn't put the computer offline.
```

I expect some regressions in plugins, but I agree it makes sense to implement the listener invocation properly. There is a `preOnline()` hook which puts agens to offline state if something goes wrong, and API users should use it anyway

**Comment by bitwiseman** [comment]
@oleg-nenashev  
The question is whether *other* exceptions aside from the two listed (such as those extending RuntimeException) should also not put the agent offline.

**Comment by oleg-nenashev** [comment]
@bitwiseman good question... my suggestion would be to ignore all exceptions like we do in other Listener implementations. ComputerListener Javadocs may need a minor update to make it explicit

**Comment by bitwiseman** [comment]
> @bitwiseman good question... my suggestion would be to ignore all exceptions like we do in other Listener implementations. ComputerListener Javadocs may need a minor update to make it explicit

Already there in this PR.  😄 

Separate point:  Not sure why compilation is failing on linux.

**Comment by oleg-nenashev** [comment]
It is an INFRA bug, already reported it to @olblak

**Review by oleg-nenashev — APPROVED** [review]
Looks good to me, and thanks for the tests

**Review by Wadeck — APPROVED** [review]
:bee: Nothing seems wrong there.

Please revert the whitespaces / format changes to ease review, thanks!

File: core/src/main/java/hudson/slaves/SlaveComputer.java [hunk 1]
```diff
@@ -694,7 +694,25 @@ public void onClosed(Channel c, IOException cause) {
         old = ACL.impersonate(ACL.SYSTEM);
         try {
             for (ComputerListener cl : ComputerListener.all()) {
-                cl.onOnline(this,taskListener);
+                try {
+                    cl.onOnline(this,taskListener);
+                } catch (IOException | InterruptedException e) {
```
Comment (jvz) [inline core/src/main/java/hudson/slaves/SlaveComputer.java]: It's typically a good idea to set the interrupt flag on the current thread after catching an interrupted exception.

File: core/src/main/java/jenkins/model/Jenkins.java [hunk 2]
```diff
@@ -979,16 +979,15 @@ protected void doRun() throws Exception {
...
-                            if (t instanceof Error) {
-                                // We propagate Runtime errors, because they are fatal.
-                                throw t;
-                            }
-
-                            // Other exceptions should be logged instead of failing the Jenkins startup (See listener's Javadoc)
-                            // We also throw it for InterruptedException since it's what is expected according to the javadoc
-                            LOGGER.log(SEVERE, String.format("Invocation of the computer listener %s failed for the Jenkins master node",
-                                    cl.getClass()), t);
+                        } catch (IOException | InterruptedException e) {
```
Comment (jvz) [inline core/src/main/java/jenkins/model/Jenkins.java]: Same here; re-interrupt.

**Review by jvz — APPROVED** [review]
Thanks for de-Eclipseing the files, too (trailing whitespace).

**Comment by oleg-nenashev** [comment]
will merge tomorrow if no negative feedback. @bitwiseman for backporting purposes, you may really want to rollback unrelated formatting changes

File: core/src/main/java/hudson/slaves/SlaveComputer.java [hunk 3]
```diff
@@ -694,7 +694,25 @@ public void onClosed(Channel c, IOException cause) {
...
+                    // Per Javadoc log exceptions but still go online.
+                    taskListener.getLogger().format(
+                        "onOnline: %s reported an exception: %s%n",
+                        cl.getClass(),
+                        e.toString());
+                } catch (Exception e) {
```
Comment (jeffret-b) [inline core/src/main/java/hudson/slaves/SlaveComputer.java]: I can't detect a difference between this catch and the prior one starting on line 699. Is there any reason not to combine them together?

File: core/src/main/java/jenkins/model/Jenkins.java [hunk 4]
```diff
@@ -979,16 +979,15 @@ protected void doRun() throws Exception {
...
+                            // Per Javadoc log exceptions but still go online.
+                            LOGGER.log(WARNING, String.format("Exception in onOnline() for the computer listener %s on the Jenkins master node",
+                                cl.getClass()), e);
+                        } catch (Exception e) {
```
Comment (jeffret-b) [inline core/src/main/java/jenkins/model/Jenkins.java]: Looks like the only difference here is the log level. I'm not sure there's an easy way to combine these and keep the different log levels, but I'm not sure the different log levels are necessary.

File: core/src/main/java/hudson/slaves/SlaveComputer.java [hunk 5]
```diff
@@ -694,7 +694,25 @@ public void onClosed(Channel c, IOException cause) {
...
+                    // Per Javadoc log other exceptions but still go online.
+                    // NOTE: this does not include Errors, which are indiciate a fatal problem
```
Comment (jeffret-b) [inline core/src/main/java/hudson/slaves/SlaveComputer.java]:
```suggestion
                    // NOTE: this does not include Errors, which indicate a fatal problem
```

File: core/src/main/java/jenkins/model/Jenkins.java [hunk 6]
```diff
@@ -979,16 +979,15 @@ protected void doRun() throws Exception {
...
+                            // Per Javadoc log other exceptions but still go online.
+                            // NOTE: this does not include Errors, which are indiciate a fatal problem
```
Comment (jeffret-b) [inline core/src/main/java/jenkins/model/Jenkins.java]:
```suggestion
                            // NOTE: this does not include Errors, which indicate a fatal problem
```

File: test/src/test/java/jenkins/model/JenkinsTest.java [hunk 7]
```diff
@@ -472,41 +475,66 @@ public void runScriptOnOfflineComputer() throws Exception {
 
     @Test
     @Issue("JENKINS-38487")
-    public void startupShouldNotFailOnFailingOnlineListener() {
-        // We do nothing, FailingOnOnlineListener & JenkinsRule should cause the 
+    public void startupShouldNotFailOnIOExceptionOnlineListener() {
+        // We do nothing, IOExceptionOnOnlineListener & JenkinsRule should cause the
         // boot failure if the issue is not fixed.
+
+        Assert.assertEquals(1, IOExceptionOnOnlineListener.onOnlineCount);
```
Comment (jeffret-b) [inline test/src/test/java/jenkins/model/JenkinsTest.java]:
```suggestion
        assertEquals(1, IOExceptionOnOnlineListener.onOnlineCount);
```
I suggest following the existing pattern in this file and calling the assert() methods directly. Then the new import also wouldn't be needed.

File: test/src/test/java/jenkins/model/JenkinsTest.java [hunk 8]
```diff
@@ -472,41 +475,66 @@ public void runScriptOnOfflineComputer() throws Exception {
...
         }
     }
-    
+
+    @Test
+    @Issue("JENKINS-57111")
+    public void startupShouldNotFailOnRuntimeExceptionOnlineListener() {
+        // We do nothing, RuntimeExceptionOnOnlineListener & JenkinsRule should cause the
+        // boot failure if the issue is not fixed.
+        Assert.assertEquals(1, RuntimeExceptionOnOnlineListener.onOnlineCount);
```
Comment (jeffret-b) [inline test/src/test/java/jenkins/model/JenkinsTest.java]:
```suggestion
        assertEquals(1, RuntimeExceptionOnOnlineListener.onOnlineCount);
```

**Review by jeffret-b — APPROVED** [review]
The change looks good and it's great to have the tests. I've left a few minor comments that I think would help consistency and clarity.

File: core/src/main/java/hudson/slaves/SlaveComputer.java (hunk 1 above)
Comment (bitwiseman) [reply core/src/main/java/hudson/slaves/SlaveComputer.java]: @jvz 
Hm, I didn't know that.  Thanks!  

In other cases I would make the change you suggested, but the point here is to log any exceptions and otherwise ignore them if at all possible. 
 
I don't think we don't want to mark the thread as interrupted here, but I'm open to being wrong.  Can you say a bit more about this?

File: core/src/main/java/hudson/slaves/SlaveComputer.java (hunk 3 above)
Comment (bitwiseman) [reply core/src/main/java/hudson/slaves/SlaveComputer.java]: I kept them separate because the API specifically names these two types.  I can unify them sure.

File: core/src/main/java/jenkins/model/Jenkins.java (hunk 4 above)
Comment (bitwiseman) [reply core/src/main/java/jenkins/model/Jenkins.java]: Good point.  The SEVERE preserve the existing behavior, but I can change them both to warning.

**Comment by bitwiseman** [comment]
Thanks for the feedback everyone!  Much higher quality due to your comments - all addressed.

File: core/src/main/java/hudson/slaves/SlaveComputer.java (hunk 1 above)
Comment (jvz) [reply core/src/main/java/hudson/slaves/SlaveComputer.java]: Basically, if you're catching that exception, it's because you were already interrupted, so any IO being cancelled was already cancelled. When you go to set `Thread.currentThread().interrupt()` a second time, it just sets the interrupt status back that was cleared upon catching the exception. Unless this is the end point of the thread at which point it doesn't matter in practice since the thread is about to finish executing.

Typical usage from the other side is something like: `if (Thread.interrupted()) throw new InterruptedException();` during a wait loop of some sort. The if check clears out the interrupted status. By not resetting the interrupt status, that could potentially lead to some sort of deadlock type scenario where, say, an interruptible call to `wait()` is never interrupted because it doesn't know the flag was set.

I originally learned this from an automatic IntelliJ code cleanup suggestion.

TL;DR: multithreading is complicated
